src/layer.rs

```diff
@@ -1,102 +1,732 @@
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
+pub struct FillLayout {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub fill_sort_key: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub visibility: Option<Visibility>,
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct BackgroundLayer {
+    pub id: String,
+    #[serde(default)]
+    pub layout: BackgroundLayout,
+    #[serde(default)]
+    pub paint: BackgroundPaint,
+}
+
+impl IntoLayer for BackgroundLayer {
+    fn into_layer(self) -> Layer {
+        Layer::Background(self)
+    }
+}
+
+impl BackgroundLayer {
+    pub fn new(id: impl Into<String>) -> BackgroundLayer {
+        BackgroundLayer {
+            id: id.into(),
+            layout: BackgroundLayout::default(),
+            paint: BackgroundPaint::default(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
+pub struct BackgroundPaint {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub background_color: Option<Expression<()>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub background_emissive_strength: Option<Expression<()>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub background_opacity: Option<Expression<()>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub background_pattern: Option<Expression<()>>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
+pub struct BackgroundLayout {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub visibility: Option<Visibility>,
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct LineLayer {
+    #[serde(flatten)]
+    pub inner: LayerBase,
+    #[serde(default)]
+    pub layout: LineLayout,
+    #[serde(default)]
+    pub paint: LinePaint,
+}
+
+impl IntoLayer for LineLayer {
+    fn into_layer(self) -> Layer {
+        Layer::Line(self)
+    }
+}
+
+impl LineLayer {
+    pub fn new(id: impl Into<String>, source: impl Into<String>) -> LineLayer {
+        LineLayer {
+            inner: LayerBase::new(id, source),
+            layout: LineLayout::default(),
+            paint: LinePaint::default(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
+pub enum LineCap {
+    #[default]
+    Butt,
+    Round,
+    Square,
+}
+impl EnumMarker for LineCap {}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
+pub enum LineJoin {
+    #[default]
+    Miter,
+    Bevel,
+    Round,
+    None,
+}
+impl EnumMarker for LineJoin {}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
```

Comment:
Feel like having [skip_serializing_none](https://docs.rs/serde_with/latest/serde_with/attr.skip_serializing_none.html) from serde_with, as all the fields are with `Option::is_none` maybe the refactoring should happen in a separate PR though